sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -173,25 +185,41 @@ private[csv] class UnivocityParser(
    */
   def parse(input: String): Option[InternalRow] = {
     convertWithParseMode(parser.parseLine(input)) { tokens =>
-      var i: Int = 0
-      while (i < indexArr.length) {
-        val pos = indexArr(i)
-        // It anyway needs to try to parse since it decides if this row is malformed
-        // or not after trying to cast in `DROPMALFORMED` mode even if the casted
-        // value is not stored in the row.
-        val value = valueConverters(pos).apply(tokens(pos))
-        if (i < requiredSchema.length) {
-          row(i) = value
+      var foundMalformed: Boolean = false
+      indexArr.foreach { case (pos, i) =>
+        try {
+          // It anyway needs to try to parse since it decides if this row is malformed
+          // or not after trying to cast in `DROPMALFORMED` mode even if the casted
+          // value is not stored in the row.
+          val value = valueConverters(pos).apply(tokens(pos))
+          if (i < requiredSchema.length) {
+            row(i) = value
+          }
+        } catch {
+          case _: NumberFormatException | _: IllegalArgumentException
```

Comment:
Could we maybe decouple this parsing handling logics into `convertWithParseMode`?